### ISSUE TITLE
error XA2006: Could not resolve reference to 'System.Void Android.Support.CustomTabs.CustomTabsIntent::LaunchUrl(Android.Content.Context,Android.Net.Uri)'

### ISSUE THREAD

**Issue by biapar** [xamarin/Xamarin.Auth#197]
I've this error on Android release mode after update to last Xamarin.Auth version

Java.Interop.Tools.Diagnostics.XamarinAndroidException: error XA2006: Could not resolve reference to 'System.Void Android.Support.CustomTabs.CustomTabsIntent::LaunchUrl(Android.Content.Context,Android.Net.Uri)' (defined in assembly 'Xamarin.Auth, Version=1.5.0.0, Culture=neutral, PublicKeyToken=null') with scope 'Xamarin.Android.Support.CustomTabs, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. When the scope is different from the defining assembly, it usually means that the type is forwarded. ---> Mono.Cecil.ResolutionException: Failed to resolve System.Void Android.Support.CustomTabs.CustomTabsIntent::LaunchUrl(Android.Content.Context,Android.Net.Uri)

How to solve?

Xamarin.Forms 2.3.4.231
Xamarin.Auth 1.5.0.3
CustomTabs 23.3.0
Android 6

**Comment by biapar** [comment]
I tried to install the CustomTabs 25.3.1 as you linked in your project, but this is the error:

Severity	Code	Description	Project	File	Line	Suppression State
Error		Could not install package 'Xamarin.Android.Support.CustomTabs 25.3.1'. You are trying to install this package into a project that targets 'MonoAndroid,Version=v6.0', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.			0

**Comment by biapar** [comment]
I updated all and the error disappear.

Latest stable Xamarin Forms
Installed Android SDK level 24,25,26
Update Android Support Libs to 25.3.x

**Comment by jenart** [comment]
The only problem with that, is that the latest Xamarin.Forms v2.3.4.247 is bound to Xamarin.Android.* libraries with v23.3.0. It would appear that it is impossible to keep your Android libs @ v23.3.0 and use Xamarin.Auth v1.5.0.3 as it now has the dependency on Xamarin.Android.* v25.3.1. 

I have tried updating the Xamarin.Android libs to v25.3.1 so I can use the sample code as shown in the Xamarin.Auth project to log into Facebook (using Native UI/Custom Renderers), but my app crashes immediately upon loading on my device (Samsung S3 GT-I9300 Android Version 4.3). So it definitely is not compatible.

The whole Social/Auth login landscape is completely fractured and confusing as far as Xamarin.Forms is concerned. There are so many different libraries to do the same thing - or so it would seem - but none other than Xamarin.Auth are suitable (or easy to implement without detailed native platform knowledge) if you are using Xamarin.Forms. 

It isn't actually even very clear at first that Xamarin.Auth is the library you should be using if you are looking to implement Native/Custom social Authentication in your app. There is no mention on the Xamarin.Auth wiki that if you use this you won't be needing Xamarin.Facebook.iOS or Xamarin.Facebook.Android packages for Facebook login (maybe I'm wrong??). There are a number of other packages all trying to do the same thing, but none are seem very suited for use with a Xamarin.Forms app. 

There are implementations for Xamarin Native (C#) iOS & Android apps, but there is no clear reference implementation on how to do it in a Xamarin.Forms app. I've managed to get Xamarin.Facebook.iOS working in my Xamarin.Forms iOS app, (See https://adrianhall.github.io/develop-mobile-apps-with-csharp-and-azure/chapter2/social/#client-flow-for-social-providers) but can't for the life of me figure out how to get Xamarin.Facebook.Android implementation working for my Android Forms app. 

Realising that Xamarin.Auth also supported Client-Flow (Native/Custon UI) for Social Auth gave me new hope, but sadly it's now been days of chasing my tail through different rabbit warrens of dependency versions, linking vs not linking, java compilation issues, and so on! Wasn't Xamarin.Forms supposed to make all this easier? 

At this point, I've just about given up on adding Android support for my app. Clearly Xamarin.Forms and any working solution for Native/Custom UI Auth is just not mature enough yet - there are too many issues just now. I will concentrate on iOS - which I have a working Auth solution for now, and will come back to Android when it has a better experience. For now, I have to keep my Xamarin.Android.* libs pinned at v23.3.0, so my UI actually works. 

**My Stack:**
* Microsoft Visual Studio Professional 2017 
* Version 15.2 (26430.12) Release VisualStudio.15.Release/15.2.0+26430.12
* Microsoft .NET Framework Version 4.6.01586

* Installed Version: Professional
* Xamarin   4.5.0.476 (1be4f0c)
* Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

* Xamarin.Android SDK   7.3.1.2 (9dbc4c5)
* Xamarin.Android Reference Assemblies and MSBuild support.

* Xamarin.iOS and Xamarin.Mac SDK   10.10.0.35 (4cffddf)
* Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

* Compile using Android v7.1 (Nougat)
* Minimum Android v4.0.3 (API Level 15 - Ice Cream Sandwich)
* Target Android v7.0.1 (API Level 25 - Nougat)

* Linking - Sdk and User Assemblies (NOTE: have tried all settings here, makes no difference. if it compiles, with no linking, it only defers the errors until runtime on the device)

**Comment by biapar** [comment]
So, Do you suggest to revert to Xamarin.Auth 1.4.1 and Xamarin.Android libs to v23.3.0?
But Xamarin.Auth 1.4.1 is gone into nuget.
I don't understand which is the stable combination.

@moljac **Could you help us and do some explanation**?

**Comment by biapar** [comment]
@jenart see here how I solve

https://github.com/xamarin/Xamarin.Auth/issues/201

**Comment by moljac** [comment]
@jenart 

> The only problem with that, is that the latest Xamarin.Forms v2.3.4.247 is bound to
> Xamarin.Android.* libraries with v23.3.0. It would appear that it is impossible to keep your Android
> libs @ v23.3.0 and use Xamarin.Auth v1.5.0.3 as it now has the dependency on Xamarin.Android.* 
> v25.3.1.

Exactly. It had to be pinned because of Xamarin.Forms. What other users do is to bump all Xamarin.Android.Support to latest versions after installing nugets and ensure that everything is the same version (in Visual Studio, Solution Node, Manage nuget packages for solution, Consolidate)


> I have tried updating the Xamarin.Android libs to v25.3.1 so I can use the sample code as shown 
> in the Xamarin.Auth project to log into Facebook (using Native UI/Custom Renderers), but my app
> crashes immediately upon loading on my device (Samsung S3 GT-I9300 Android Version 4.3). 
> So it definitely is not compatible.

This is most likely Samsung specific problem (searching for packages supporting CustomTabs). Can you test on other devices and/or provide minimal sample for triaging?

> The whole Social/Auth login landscape is completely fractured and confusing as far as 
> Xamarin.Forms is concerned. There are so many different libraries to do the same thing - or so it 
 > would seem - but none other than Xamarin.Auth are suitable (or easy to implement without 
> detailed native platform knowledge) if you are using Xamarin.Forms.

I added Xamarin.Forms under pressure. I was not able to even answer "how to do Xamarin.Auth with Xamarin.Forms". Presenters implementation should be quite stable, while CustomRenderers are not extensively tested.

> t isn't actually even very clear at first that Xamarin.Auth is the library you should be using if you 
> are looking to implement Native/Custom social Authentication in your app. 

I simply cannot make everything on time. Wiki is heave construction site.
I think it is quite clearly said it can be used with Custom Tabs and SafariWebView, but let me know where I should put it.

> There is no mention on the Xamarin.Auth wiki that if you use this you won't be needing
> Xamarin.Facebook.iOS or Xamarin.Facebook.Android packages for Facebook login (maybe I'm
> wrong??). 

So under native you understand Native SDKs by Google and Facebook? OK. Xamarin.Auth is no replacement for those and you are not wrong, but there will be no advertisements that Xamarin.Auth replaces Native SDKs. Xamarin.Auth will help you get token and authenticate and do few other things, but not to replace Native SDKs. There are plans to add wrappers for those, so Xamarin.Auth would call native SDK, but this is on wish list.

> There are a number of other packages all trying to do the same thing, but none are seem very
> suited for use with a Xamarin.Forms app.

Check those if they suite you:

https://github.com/Redth/Plugin.SocialAuth
https://github.com/Clancey/SimpleAuth

> There are implementations for Xamarin Native (C#) iOS & Android apps, but there is no clear
> reference implementation on how to do it in a Xamarin.Forms app. I've managed to get
>  Xamarin.Facebook.iOS working in my Xamarin.Forms iOS app, (See
> https://adrianhall.github.io/develop-mobile-apps-with-csharp-and-azure/chapter2/social/#client-flow-for-social-providers) 
> but can't for the life of me figure out how to get Xamarin.Facebook.Android implementation
> working for my Android Forms app.

Due to lack of time I cannot speak about other libs (did not test them). Sorry.

> Realising that Xamarin.Auth also supported Client-Flow (Native/Custon UI) for Social Auth gave 
> me new hope, but sadly it's now been days of chasing my tail through different rabbit warrens of 
> dependency versions, linking vs not linking, java compilation issues, and so on! Wasn't 
> Xamarin.Forms supposed to make all this easier?

Yes, but the Android landscape is complex itself add to that Xamarin Standard/Traditional and another abstraction on top of it. You should get the picture.

> At this point, I've just about given up on adding Android support for my app. Clearly 
> Xamarin.Forms and any working solution for Native/Custom UI Auth is just not mature enough yet > - there are too many issues just now. I will concentrate on iOS - which I have a working Auth
> solution for now, and will come back to Android when it has a better experience. For now, I have to
> keep my Xamarin.Android.* libs pinned at v23.3.0, so my UI actually works.

The biggest issue is - package management and getting everything right. The most problems arose with Xamarin.Forms support, but it was expected.

Xamarin.Auth community (and myself) hangs out in Community slack channel 24x7 and they take burden off my back by helping in real time other (newer) users. I'd suggest you to go there, so we can help you solve those issues

**Comment by jenart** [comment]
@moljac thanks for taking the time to write your response. I appreciate your time.

I've updated all the Xamarin.Android.Support libraries to the latest versions. I changed the Android project settings to target Android 6.0 for compilation and set the minimum version to Android v6.0 as well. I deployed to the Android Emulator running x86 with API 23 Android 6.0 and the application still doesn't work. No button click/tap events are firing. 

My main entry form (XAML Page) loads with 3 buttons and previously with Xamarin.Forms v23.3.0 (prior to attempting to use Xamarin.Auth) they all worked perfectly on both the device (Samsung S3) and the same Android 6.0 x86 Emulator. So my apps functionality is actually going backwards at this point. I'm going to put Android/Xamarin.Forms social login functionality on hold for now and come back and retry it all at a later date.

I really don't understand why Xamarin.Auth (v1.5.0.3) has chosen to take a dependency on Xamarin.Android.Support library functionality (v25.3.1). This is not compatible with the Xamarin.Android.Support libraries (v23.3.0) that the current Xamarin.Forms (v2.3.4.247) release is build against.

**Comment by moljac** [comment]
@jenart 

> thanks for taking the time to write your response. I appreciate your time.

You are welcome.

> I've updated all the Xamarin.Android.Support libraries to the latest versions. I changed the Android project settings to target Android 6.0 for compilation and set the minimum version to Android v6.0 as well. I deployed to the Android Emulator running x86 with API 23 Android 6.0 and the  application still doesn't work. No button click/tap events are firing.

Can you try 7.x?

> My main entry form (XAML Page) loads with 3 buttons and previously with Xamarin.Forms v23.3.0 (prior to attempting to use Xamarin.Auth) they all worked perfectly on both the device (Samsung S3) and the same Android 6.0 x86 Emulator. So my apps functionality is actually going backwards at this point. I'm going to put Android/Xamarin.Forms social login functionality on hold for now and come back and retry it all at a later date.

Sorry to hear that, but this seems like not Xamarin.Auth problem - buttons not working?!?

> I really don't understand why Xamarin.Auth (v1.5.0.3) has chosen to take a dependency on Xamarin.Android.Support library functionality (v25.3.1). This is not compatible with the Xamarin.Android.Support libraries (v23.3.0) that the current Xamarin.Forms (v2.3.4.247) release is build against.

I will gather info about 23.3.0. It is pinned to this version only because of Forms support. I (and most of other users) update Support nugets to latest and it works in most cases.

If you stay on 23.3.0 you get error mentioned in the title of this issue.

**Comment by jenart** [comment]
@moljac  It seems like this post by @jamesmontemagno is pretty clear about the need to stay with the pinned Xamarin.Android.Support.* libraries and the current Xamarin.Forms version:

https://forums.xamarin.com/discussion/57283/unable-to-find-a-version-of-xf-compatible-with

With this in mind, I really don't want to be trying go against this advice. I don't know Android development well at all (which is why I chose Xamarin.Forms to abstract it all away) so I don't want to have to be chasing down obscure compatibility issues/bugs because I took a higher dependency than the Xamarin guys advise!

Also, a major consideration that I have for my app is that I want to cover the largest percentage of android phones out in the wild:

https://developer.android.com/about/dashboards/index.html
https://www.statista.com/statistics/271774/share-of-android-platforms-on-mobile-devices-with-android-os/

Based on the usage statistics listed on these sites you can get approximately 80% device coverage by targeting Android v4.4 (KitKat) - which still has nearly 20% of the marketplace!! I don't know if I'm ready to give up such a huge potential target market by bumping the target version to Android v6.0 or higher!

**Comment by jamesmontemagno** [comment]
@jenart  watch this:
https://channel9.msdn.com/Shows/XamarinShow/Snack-Pack-11-Understanding-Android-API-Level-Settings

or read this:
https://developer.xamarin.com/guides/android/application_fundamentals/understanding_android_api_levels/

You aren't targeting that version... minimum is what says how far back you can go.

**Comment by biapar** [comment]
A good free Xamarin Blog post about Android support lib
https://blog.xamarin.com/mastering-android-support-libraries/

**Comment by sjreed** [comment]
I have been working with Xamarin.Auth for several months in DEBUG mode while building my Xamarin Forms app.  I now switch to RELEASE mode and it fails with the linkage error described in this post title.

My Xamarin Forms is 2.3.4.247, which requires support modules of 23.3.0.

I don't understand how I can use Xamarin.Auth 1.5.3 when it seems to point to 25.3.1 libraries!

**Comment by jamesmontemagno** [comment]
@sjreed  Xamarin.Forms has not required 23.3.0 for a long time. you just need to adjust your compile target: https://blog.xamarin.com/mastering-android-support-libraries/

**Comment by sjreed** [comment]
You are correct in stating that Xamarin.Forms does not require 23.3.0, as long as you can force it to switch to Mono.Android = v7.0.  Your article clarified how to get that done.  Thank you for that!

After completing that process and  restarting VS a couple of times I was successful in getting a clean debug compile.  Then it was a matter of fighting other battles with Java Max Heap Size and ProGuard to actually get a release compile created.

Bottom line, your answer solved my problem!  Thanks!

**Comment by newky2k** [comment]
resolved
